Dgraph v25.1.0 alpha Crashing

## Describe the bug

Hi team, we're running dgraph v25.1.0 in kubenetes,  3 alpha nodes. we noticed that all the 3 alpha nodes crashed almost at the same time. the following logs are from the crash alpha pod.  this happend twice in 10 days. please help take a look, thanks in advance.

Error log:
alpha-0:
```
I0330 08:23:03.856631       1 groups.go:153] Informed Zero about tablets I have: OK
I0330 08:23:03.856747       1 groups.go:156] Upserted Schema and Types: OK
I0330 08:23:03.856759       1 groups.go:159] Server is ready: OK
I0330 08:23:03.856770       1 access.go:419] InitializeAcl closed
I0330 08:23:04.575569       1 log.go:22] 1 [term: 43] received a MsgVote message with higher term from 2 [term: 44]
I0330 08:23:04.575625       1 log.go:22] 1 became follower at term 44
I0330 08:23:04.575646       1 log.go:22] 1 [logterm: 43, index: 471104830, vote: 0] cast MsgVote for 2 [logterm: 43, index: 471104830] at term 44
I0330 08:23:04.575908       1 log.go:22] 1 [logterm: 43, index: 471104830, vote: 2] rejected MsgVote from 3 [logterm: 43, index: 471104830] at term 44
I0330 08:23:05.577528       1 log.go:22] raft.node: 1 elected leader 2 at term 44
I0330 08:23:06.318047       1 access.go:354] RefreshAcls closed
I0330 08:23:07.619510       1 admin.go:846] namespace: 0. No GraphQL schema in Dgraph; serving empty GraphQL API
I0330 08:23:44.652189       1 draft.go:773] Creating snapshot at Index: 471105243, ReadTs: 807437427
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1d8ae12]

goroutine 26443 [running]:
github.com/dgraph-io/dgraph/v25/dql.TypeValFrom(0xc00bc6f2c0?)
        /home/runner/_work/dgraph/dgraph/dql/mutation.go:53 +0x12
github.com/dgraph-io/dgraph/v25/edgraph.verifyUniqueWithinMutation(0xc00ba806c0)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:2162 +0x10a
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).doMutate(0x292f868?, {0x292f868, 0xc04eb7f110}, 0xc00ba806c0, 0xc00b571180)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:542 +0x11b
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).doQuery(0x3b81580, {0x292f868, 0xc04ec98c60}, 0xc008bd3710)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:1367 +0xf89
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).QueryNoGrpc(0x3b81580, {0x292f868?, 0xc04eb7eff0?}, 0xc00b870d00)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:1217 +0x2b1
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).Query(0x7f4eb01042b0?, {0x292f868, 0xc04eb7eff0}, 0xc0000b7008?)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:1183 +0x2f
github.com/dgraph-io/dgo/v250/protos/api._Dgraph_Query_Handler.func1({0x292f868?, 0xc04eb7eff0?}, {0x2569240?, 0xc00b870d00?})
        /home/runner/go/pkg/mod/github.com/dgraph-io/dgo/v250@v250.0.0/protos/api/api_grpc.pb.go:320 +0xcb
github.com/dgraph-io/dgraph/v25/audit.AuditRequestGRPC({0x292f868?, 0xc04eb7eff0?}, {0x2569240?, 0xc00b870d00?}, 0xd31225?, 0x80?)
        /home/runner/_work/dgraph/dgraph/audit/interceptor.go:72 +0xe2
github.com/dgraph-io/dgo/v250/protos/api._Dgraph_Query_Handler({0x25b26e0, 0x3b81580}, {0x292f868, 0xc04eb7eff0}, 0xc00b870c80, 0x26e6f68)
        /home/runner/go/pkg/mod/github.com/dgraph-io/dgo/v250@v250.0.0/protos/api/api_grpc.pb.go:322 +0x143
google.golang.org/grpc.(*Server).processUnaryRPC(0xc006954000, {0x292f868, 0xc04eb7eed0}, 0xc00b84f740, 0xc0000b4390, 0x3718a38, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1431 +0x1036
google.golang.org/grpc.(*Server).handleStream(0xc006954000, {0x2930ab8, 0xc00ba98000}, 0xc00b84f740)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1842 +0xb88
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 26442
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x11d
```
alpha-1 crash log
```
I0329 07:10:08.473473       1 draft.go:773] Creating snapshot at Index: 451538606, ReadTs: 767099652
I0329 07:17:08.390412       1 draft.go:773] Creating snapshot at Index: 451703491, ReadTs: 767395536
I0329 07:22:08.406427       1 draft.go:773] Creating snapshot at Index: 451837821, ReadTs: 767636562
I0329 07:28:08.385911       1 draft.go:773] Creating snapshot at Index: 451992040, ReadTs: 767912832
I0329 07:34:08.552871       1 draft.go:773] Creating snapshot at Index: 452152138, ReadTs: 768205636
I0329 07:39:08.420615       1 draft.go:773] Creating snapshot at Index: 452287570, ReadTs: 768450112
I0329 07:45:08.382375       1 draft.go:773] Creating snapshot at Index: 452442424, ReadTs: 768728589
I0329 07:51:08.341808       1 draft.go:773] Creating snapshot at Index: 452592096, ReadTs: 768997511
I0329 07:57:08.541986       1 draft.go:773] Creating snapshot at Index: 452749964, ReadTs: 769281157
I0329 08:03:08.452859       1 draft.go:773] Creating snapshot at Index: 452907172, ReadTs: 769565445
I0329 08:09:08.341874       1 draft.go:773] Creating snapshot at Index: 453046204, ReadTs: 769816117
I0329 08:17:08.434087       1 draft.go:773] Creating snapshot at Index: 453192287, ReadTs: 770102680
I0329 08:23:08.549832       1 draft.go:773] Creating snapshot at Index: 453357310, ReadTs: 770452238
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1d8ae12]

goroutine 18546320 [running]:
github.com/dgraph-io/dgraph/v25/dql.TypeValFrom(0xc07563fd60?)
        /home/runner/_work/dgraph/dgraph/dql/mutation.go:53 +0x12
github.com/dgraph-io/dgraph/v25/edgraph.verifyUniqueWithinMutation(0xc070195ec0)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:2162 +0x10a
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).doMutate(0x292f868?, {0x292f868, 0xc0824812f0}, 0xc070195ec0, 0xc06f16dc00)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:542 +0x11b
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).doQuery(0x3b81580, {0x292f868, 0xc081835260}, 0xc06ede5710)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:1367 +0xf89
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).QueryNoGrpc(0x3b81580, {0x292f868?, 0xc0824811d0?}, 0xc06f16d780)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:1217 +0x2b1
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).Query(0x7fe0540adcb8?, {0x292f868, 0xc0824811d0}, 0xc076200008?)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:1183 +0x2f
github.com/dgraph-io/dgo/v250/protos/api._Dgraph_Query_Handler.func1({0x292f868?, 0xc0824811d0?}, {0x2569240?, 0xc06f16d780?})
        /home/runner/go/pkg/mod/github.com/dgraph-io/dgo/v250@v250.0.0/protos/api/api_grpc.pb.go:320 +0xcb
github.com/dgraph-io/dgraph/v25/audit.AuditRequestGRPC({0x292f868?, 0xc0824811d0?}, {0x2569240?, 0xc06f16d780?}, 0xd31225?, 0x80?)
        /home/runner/_work/dgraph/dgraph/audit/interceptor.go:72 +0xe2
github.com/dgraph-io/dgo/v250/protos/api._Dgraph_Query_Handler({0x25b26e0, 0x3b81580}, {0x292f868, 0xc0824811d0}, 0xc06f16d700, 0x26e6f68)
        /home/runner/go/pkg/mod/github.com/dgraph-io/dgo/v250@v250.0.0/protos/api/api_grpc.pb.go:322 +0x143
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000b34000, {0x292f868, 0xc0824810b0}, 0xc0701cc120, 0xc00ea0fbc0, 0x3718a38, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1431 +0x1036
google.golang.org/grpc.(*Server).handleStream(0xc000b34000, {0x2930ab8, 0xc002f7ba00}, 0xc0701cc120)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1842 +0xb88
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 18136657
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x11d
```
alpha-2 crash log
```
I0330 08:24:28.339065       1 pool.go:321] CONN: Re-established connection with dgraph-v25-main-alpha-0.dgraph-v25-main-alpha-headless.catalog.svc.cluster.local:7080.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1d8ae12]

goroutine 36964037 [running]:
github.com/dgraph-io/dgraph/v25/dql.TypeValFrom(0xc06f94ae60?)
        /home/runner/_work/dgraph/dgraph/dql/mutation.go:53 +0x12
github.com/dgraph-io/dgraph/v25/edgraph.verifyUniqueWithinMutation(0xc05a30af00)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:2162 +0x10a
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).doMutate(0x292f868?, {0x292f868, 0xc0776be4b0}, 0xc05a30af00, 0xc059234c00)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:542 +0x11b
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).doQuery(0x3b81580, {0x292f868, 0xc074afa7b0}, 0xc056e57710)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:1367 +0xf89
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).QueryNoGrpc(0x3b81580, {0x292f868?, 0xc0776be2a0?}, 0xc05a31e100)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:1217 +0x2b1
github.com/dgraph-io/dgraph/v25/edgraph.(*Server).Query(0x7f8b340d8a80?, {0x292f868, 0xc0776be2a0}, 0xc048d80008?)
        /home/runner/_work/dgraph/dgraph/edgraph/server.go:1183 +0x2f
github.com/dgraph-io/dgo/v250/protos/api._Dgraph_Query_Handler.func1({0x292f868?, 0xc0776be2a0?}, {0x2569240?, 0xc05a31e100?})
        /home/runner/go/pkg/mod/github.com/dgraph-io/dgo/v250@v250.0.0/protos/api/api_grpc.pb.go:320 +0xcb
github.com/dgraph-io/dgraph/v25/audit.AuditRequestGRPC({0x292f868?, 0xc0776be2a0?}, {0x2569240?, 0xc05a31e100?}, 0xd31225?, 0x80?)
        /home/runner/_work/dgraph/dgraph/audit/interceptor.go:72 +0xe2
github.com/dgraph-io/dgo/v250/protos/api._Dgraph_Query_Handler({0x25b26e0, 0x3b81580}, {0x292f868, 0xc0776be2a0}, 0xc05a31e080, 0x26e6f68)
        /home/runner/go/pkg/mod/github.com/dgraph-io/dgo/v250@v250.0.0/protos/api/api_grpc.pb.go:322 +0x143
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000f70200, {0x292f868, 0xc0776be0c0}, 0xc05a3088a0, 0xc006d10240, 0x3718a38, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1431 +0x1036
google.golang.org/grpc.(*Server).handleStream(0xc000f70200, {0x2930ab8, 0xc04930b520}, 0xc05a3088a0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1842 +0xb88
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 36501114
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x11d
```
## To Reproduce
None


## Environment
- Kubenetes
- docker image: docker.io/dgraph/dgraph:v25.1.0
- Version [v25.1.0]

--- a/edgraph/server.go
+++ b/edgraph/server.go
@@ -1804,6 +1804,9 @@ func addQueryIfUnique(qctx context.Context, qc *queryContext) error {
 			// in the mutation, then we reject the mutation.
 
 			if !strings.HasPrefix(pred.ObjectId, "val(") {
+				if pred.ObjectValue == nil {
+					continue
+				}
 				val := strconv.Quote(fmt.Sprintf("%v", dql.TypeValFrom(pred.ObjectValue).Value))
 				query := fmt.Sprintf(`%v as var(func: eq(%v,"%v"))`, queryVar, predicateName, val[1:len(val)-1])
 				if _, err := buildQuery.WriteString(query); err != nil {
@@ -2218,6 +2221,9 @@ func verifyUniqueWithinMutation(qc *queryContext) error {
 			continue
 		}
 		pred1 := qc.gmuList[gmuIndex].Set[rdfIndex]
+		if pred1.ObjectValue == nil {
+			continue
+		}
 		pred1Value := dql.TypeValFrom(pred1.ObjectValue).Value
 		for j := range qc.uniqueVars {
 			if i == j {
@@ -2229,6 +2235,9 @@ func verifyUniqueWithinMutation(qc *queryContext) error {
 				continue
 			}
 			pred2 := qc.gmuList[gmuIndex2].Set[rdfIndex2]
+			if pred2.ObjectValue == nil {
+				continue
+			}
 			if pred2.Predicate == pred1.Predicate && dql.TypeValFrom(pred2.ObjectValue).Value == pred1Value &&
 				pred2.Subject != pred1.Subject {
 				return errors.Errorf("could not insert duplicate value [%v] for predicate [%v]",

--- a/edgraph/server_test.go
+++ b/edgraph/server_test.go
@@ -285,4 +285,94 @@ func TestVerifyUniqueWithinMutationBoundsChecks(t *testing.T) {
 		err := verifyUniqueWithinMutation(qc)
 		require.NoError(t, err)
 	})
+
+	// Regression test for https://github.com/dgraph-io/dgraph/issues/9670
+	// When an NQuad has a nil ObjectValue (e.g. a UID edge on a predicate
+	// marked @unique), verifyUniqueWithinMutation would panic with a nil
+	// pointer dereference in dql.TypeValFrom.
+	t.Run("nil ObjectValue should not panic", func(t *testing.T) {
+		qc := &queryContext{
+			gmuList: []*dql.Mutation{
+				{
+					Set: []*api.NQuad{
+						{
+							Subject:     "_:a",
+							Predicate:   "friend",
+							ObjectId:    "0x1",
+							ObjectValue: nil, // UID edge, no value
+						},
+					},
+				},
+			},
+			uniqueVars: map[uint64]uniquePredMeta{
+				encodeIndex(0, 0): {},
+			},
+		}
+		// Before fix: panics with "nil pointer dereference" in dql.TypeValFrom
+		require.NotPanics(t, func() {
+			_ = verifyUniqueWithinMutation(qc)
+		})
+	})
+
+	// Same issue but with two NQuads where one has a nil ObjectValue
+	// and the other has a real value — the inner loop comparison must
+	// also handle the nil case.
+	t.Run("nil ObjectValue mixed with value NQuad should not panic", func(t *testing.T) {
+		qc := &queryContext{
+			gmuList: []*dql.Mutation{
+				{
+					Set: []*api.NQuad{
+						{
+							Subject:   "_:a",
+							Predicate: "email",
+							ObjectValue: &api.Value{
+								Val: &api.Value_StrVal{StrVal: "test@example.com"},
+							},
+						},
+						{
+							Subject:     "_:a",
+							Predicate:   "friend",
+							ObjectId:    "0x2",
+							ObjectValue: nil, // UID edge
+						},
+					},
+				},
+			},
+			uniqueVars: map[uint64]uniquePredMeta{
+				encodeIndex(0, 0): {},
+				encodeIndex(0, 1): {},
+			},
+		}
+		require.NotPanics(t, func() {
+			err := verifyUniqueWithinMutation(qc)
+			require.NoError(t, err)
+		})
+	})
+
+	// Verify that val(...) reference edges (ObjectId="val(x)", ObjectValue=nil)
+	// don't panic. These go through a different code path in addQueryIfUnique
+	// but could still appear in uniqueVars and be iterated by
+	// verifyUniqueWithinMutation.
+	t.Run("val() reference with nil ObjectValue should not panic", func(t *testing.T) {
+		qc := &queryContext{
+			gmuList: []*dql.Mutation{
+				{
+					Set: []*api.NQuad{
+						{
+							Subject:     "_:a",
+							Predicate:   "email",
+							ObjectId:    "val(queryVar)",
+							ObjectValue: nil,
+						},
+					},
+				},
+			},
+			uniqueVars: map[uint64]uniquePredMeta{
+				encodeIndex(0, 0): {},
+			},
+		}
+		require.NotPanics(t, func() {
+			_ = verifyUniqueWithinMutation(qc)
+		})
+	})
 }

